kong/router.lua

```diff
@@ -1614,243 +1796,69 @@ function _M.new(routes)
-                        upstream_uri = upstream_base .. "/"
-                      else
-                        upstream_uri = upstream_base
-                      end
-                    elseif byte(request_postfix, 1, 1) == SLASH then
-                      upstream_uri = upstream_base .. request_postfix
-                    else
-                      upstream_uri = upstream_base .. "/" .. request_postfix
-                    end
-
-                  else -- does not end with / and strip_uri = false
-                    if req_uri == "/" then
-                      upstream_uri = upstream_base
-                    else
-                      upstream_uri = upstream_base .. req_uri
-                    end
-                  end
-                end
-              end
-
-              -- preserve_host header logic
-
-              if matched_route.preserve_host then
-                upstream_host = raw_req_host or var.http_host
-              end
-            end
-
-            local match_t     = {
-              route           = matched_route.route,
-              service         = matched_route.service,
-              headers         = matched_route.headers,
-              upstream_url_t  = upstream_url_t,
-              upstream_scheme = upstream_url_t.scheme,
-              upstream_uri    = upstream_uri,
-              upstream_host   = upstream_host,
-              prefix          = request_prefix,
-              matches         = {
-                uri_captures  = matches.uri_captures,
-                uri           = matches.uri,
-                host          = matches.host,
-                headers       = matches.headers,
-                method        = matches.method,
-                src_ip        = matches.src_ip,
-                src_port      = matches.src_port,
-                dst_ip        = matches.dst_ip,
-                dst_port      = matches.dst_port,
-                sni           = matches.sni,
-              }
-            }
-
-            if band(matched_route.match_rules, MATCH_RULES.HEADER) == 0 then
-              cache:set(cache_key, match_t)
-            end
-
-            return match_t
-          end
-        end
-
-        -- check lower category
-        category_idx = category_idx + 1
+      ctx.req_category             = req_category
+      ctx.raw_req_host             = raw_req_host
+      ctx.req_method               = req_method
+      ctx.req_uri                  = req_uri
+      ctx.req_host                 = req_host
+      ctx.req_scheme               = req_scheme
+      ctx.req_headers              = req_headers
+      ctx.src_ip                   = src_ip
+      ctx.src_port                 = src_port
+      ctx.dst_ip                   = dst_ip
+      ctx.dst_port                 = dst_port
+      ctx.sni                      = sni
+      ctx.host_with_port           = host_with_port
+      ctx.host_no_port             = host_no_port
+
+      local match_t = find_match(ctx)
```

Comment:
I think `find_match` doesn't make the code any more clear (it moves a lot of code into a function, yes, but that function has a lot of dependencies thrown into `ctx`, which makes the code more difficult to read).

I also think that it will make things slightly slower:
* Initializing `ctx` takes longer (more assignments)
* Variables that previously were accesible through locals (req_category) need to be accessed via `ctx` (`ctx.req_category`).

Reply:
On that note: have you considered removing ctx completely? Just passing the functions the parameters they need.

Reply:
The ctx is already used everywhere in the matching. So I just stuffed some vars in it. Yes there are assignments, but no table resizes (nor request time table creation), as the ctx is worker level variable and already contains these fields.

These were already in original code:
https://github.com/Kong/kong/blob/master/kong/router.lua#L1496-L1505

It is just now moved a way later. I would not like a function that takes 20 arguments or so.